src/tracks/ui/TextEditHelper.cpp

```diff
@@ -0,0 +1,666 @@
+                view.PasteSelectedText(project, newSel.t0(), newSel.t1());
+                wxTheClipboard->UsePrimarySelection(false);
+}
+#endif
+            return true;
+        }
+    }
+    return false;
+}
+
+bool TextEditHelper::OnDrag(const wxMouseEvent& event, AudacityProject* project)
+{
+    return HandleDragRelease(event, project);
+}
+
+bool TextEditHelper::OnRelease(const wxMouseEvent& event, AudacityProject* project)
+{
+    return HandleDragRelease(event, project);
+}
+
+void TextEditHelper::Draw(wxDC& dc, const wxRect& rect)
+{
+    mBBox = rect;
+    dc.SetFont(mFont);
+
+    const auto cursorHeight = dc.GetFontMetrics().height;
+
+    wxRect prevClippingRegion;
+    auto restoreClippingRegion = dc.GetClippingBox(prevClippingRegion);
+    dc.SetClippingRegion(rect);
+
+    
+    auto rtl = wxTheApp->GetLayoutDirection() == wxLayout_RightToLeft;
+
+    auto curPosX = 0;
+    auto maxOffset = static_cast<int>(mText.Length());
+    mOffset = std::clamp(mOffset, 0, maxOffset);
+    {
+        auto leftBound = rect.GetLeft();
+        auto rightBound = rect.GetRight() + 1;
+        GetCharPositionX(mCurrentCursorPos, &curPosX);
+
+        if ((!rtl && curPosX >= rightBound) || (rtl && curPosX < leftBound))
+        {
+            while (mOffset < maxOffset)
+            {
+                GetCharPositionX(mCurrentCursorPos, &curPosX);
+                if (curPosX < rightBound && curPosX >= leftBound)
+                    break;
+                ++mOffset;
+            }
+        }
+        if ((!rtl && curPosX < leftBound) || (rtl && curPosX >= rightBound))
+        {
+            while (mOffset > 0)
+            {
+                GetCharPositionX(mCurrentCursorPos, &curPosX);
+                if (curPosX >= leftBound && curPosX < rightBound)
+                    break;
+                --mOffset;
+            }
+        }
+    }
+
+    if (mCurrentCursorPos != mInitialCursorPos)
+    {
+        auto left = 0;
+        auto right = 0;
+        GetCharPositionX(std::min(mCurrentCursorPos, mInitialCursorPos), &left);
+        GetCharPositionX(std::max(mCurrentCursorPos, mInitialCursorPos), &right);
+        dc.SetPen(*wxTRANSPARENT_PEN);
+        dc.SetBrush(mTextSelectionColor);
+        dc.DrawRectangle(wxRect(left, rect.GetTop() + (rect.GetHeight() - cursorHeight) / 2, right - left, cursorHeight));
+    }
+
+    dc.SetTextBackground(wxTransparentColour);
+    dc.SetTextForeground(mTextColor);
+    dc.SetFont(wxFont(wxFontInfo()));
+    dc.DrawLabel(mText.Mid(mOffset), rect, (rtl ? wxALIGN_RIGHT : wxALIGN_LEFT) | wxALIGN_CENTER_VERTICAL);
+
+    if (mCurrentCursorPos == mInitialCursorPos)
+    {
+        dc.SetPen(mTextColor);
+        auto top = rect.GetTop() + (rect.GetHeight() - cursorHeight) / 2;
+        dc.DrawLine(curPosX, top, curPosX, top + cursorHeight);
+    }
+
+    dc.DestroyClippingRegion();
+    if (restoreClippingRegion)
+        dc.SetClippingRegion(prevClippingRegion);
```

Comment:
My preference is to put cleanup lines like the above three inside
```
auto cleanup = finally([&]{ ... });
```
And put the declaration of cleanup higher, just after the first SetClippingRegion().

This is ad-hoc RAII.  `finally` generates a class that performs the lambda in its destructor, and binds an instance to `cleanup`.

This makes it easier to be confident that a long function isn't making a mistake somewhere, with an early exit that forgets to restore the context.  Or, in case of exceptions.

Reply:
I'm not sure that undoing clipping region in the case of exception would be a safe operation

Reply:
Fixed with wxDCClipper